eclipse-theia/theia
# Delegate `ViewContainer` toolbar checks to child widgets

#### What it does

Fixes #8970 by changing the way toolbar item visibility is handled for `ViewContainers`. Instead of registering temporary toolbar items that refer to itself, a `ViewContainer` now implements the interface `TabBarDelegator`, which allows the `TabBarToolbar` system to access the child rather than the `ViewContainer`.

Paired with #8967 (and possible #9080), it should fix most of our toolbar update issues.

I've done it as two commits because giving the `TabBarToolbar` access to the `TabBarToolbarRegistry` required shifting the order in which the two things were declared in the file. The first commit contains no meaningful changes, and I didn't want to obscure the changes in the second commit by including the cut-paste as part of it.

#### How to test

1. Check that the existing toolbar items for the `scm` package behave as expected (list vs. tree toggle, collapse all). They have been adjusted to refer to the `ScmWidget` directly rather than checking the `ScmViewContainer` to find the right widget - that was a workaround to avoid #8970.
2. Check existing items for the Extensions view. They have been adjusted to be registered inside the `ViewContainer` rather than in the contribution.
3. Check that for any `ViewContainer` with multiple children, a `...` item exists that reveals a context menu that can hide and unhide children. In the Extensions view, they should appear under a `Views` submenu.
4. Check that hiding and revealing children (e.g. in the `navigator` if the NPM view is available), using that menu or the `ViewContainer`'s context menu, puts the correct items in the top bar
  - When a single child is visible, its toolbar items should populate the `ViewContainer`'s toolbar.
  - When multiple children are visible, only the `...` should be visible.


> NB: It was formerly possible to register toolbar items whose visibility depended on a check for the `ViewContainer` rather than a child widget, as was done in the `scm` package. Such checks don't appear to be possible in VSCode - toolbar items are always associated with a particular child - and they will no longer behave as expected if this PR is accepted.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
